pkg/vm/engine/tae/catalog/checkpoint.go

```diff
@@ -47,86 +47,29 @@ func CheckpointSelectOp(entry *BaseEntry, minTs, maxTs types.TS) bool {
 	if entry.InTxnOrRollbacked() {
 		return false
 	}
-	// 1. entry was created after maxTs. Skip it
-	if entry.CreateAfter(maxTs) {
-		return false
-	}
-	// 2. entry was deleted before minTs. Skip it
-	if entry.DeleteBefore(minTs) {
-		return false
-	}
-	// 3. entry was created in (0, minTs)
-	if entry.CreateBefore(minTs) {
-		// 3.1 entry was not deleted. Skip it
-		if !entry.HasDropped() {
-			return false
-		}
-		// 3.2 entry was deleted in (maxTs, +inf). Skip it
-		if entry.DeleteAfter(maxTs) {
-			return false
-		}
-	}
-	return true
+	return entry.ExistUpdate(minTs, maxTs)
+}
+
+type CatalogEntry interface {
+	GetCheckpointItems(start, end types.TS) CheckpointItems
+	RLock()
+	RUnlock()
+}
+
+type CheckpointItems interface {
+	GetIndexes() []*wal.Index
+	MakeLogEntry() *EntryCommand
 }
 
-func CheckpointOp(ckpEntry *CheckpointEntry, entry *BaseEntry, item CheckpointItem, minTs, maxTs types.TS) {
+func CheckpointOp(ckpEntry *CheckpointEntry, entry CatalogEntry, minTs, maxTs types.TS) {
 	entry.RLock()
-	if entry.InTxnOrRollbacked() {
-		entry.RUnlock()
-		return
-	}
-	// 1. entry was created in (maxTs, +inf). Skip it
-	if entry.CreateAfter(maxTs) {
-		entry.RUnlock()
-		return
-	}
-	// 2. entry was deleted in (0, minTs). Skip it
-	if entry.DeleteBefore(minTs) {
-		entry.RUnlock()
-		return
-	}
-	// 3. entry was created in (0, minTs)
-	if entry.CreateBefore(minTs) {
-		// 3.1 entry was not deleted. skip it
-		if !entry.HasDropped() {
-			entry.RUnlock()
-			return
-		}
-		// 3.2 entry was deleted (maxTs, inf). Skip it
-		if entry.DeleteAfter(maxTs) {
-			entry.RUnlock()
-			return
-		}
-		// 3.3 entry was deleted in [minTs, maxTs]
-		ckpEntry.AddIndex(entry.LogIndex)
-		cloned := item.Clone()
-		entry.RUnlock()
-		ckpEntry.AddCommand(cloned.MakeLogEntry())
-		return
-	}
-	// 4. entry was created at|after minTs
-	// 4.1 entry was deleted at|before maxTs
-	if entry.DeleteBefore(maxTs.Next()) {
-		ckpEntry.AddIndex(entry.LogIndex)
-		ckpEntry.AddIndex(entry.PrevCommit.LogIndex)
-		cloned := item.Clone()
-		entry.RUnlock()
-		ckpEntry.AddCommand(cloned.MakeLogEntry())
-		return
-	}
-	// 4.2 entry was not deleted
-	if !entry.HasDropped() {
-		ckpEntry.AddIndex(entry.LogIndex)
-		cloned := item.Clone()
-		entry.RUnlock()
-		ckpEntry.AddCommand(cloned.MakeLogEntry())
+	ckpItem := entry.GetCheckpointItems(minTs, maxTs)
```

Comment:
Again, here the lock/unlock and what the lock protects, are all very confusing.    I don't think it is right.